### PR TITLE
Remove `quic` crate feature

### LINKED ISSUE #1437: reducing Cargo feature list
Right now we have two features that control dangerous code: `dangerous_configuration` and `secret_extraction`. We also have started using the cool `--feature-powerset` flag to test various combinations of features:

```
cargo hack check --feature-powerset
```

The size of the powerset increases very fast with number of features, so it's useful to keep that number under control.

I think `dangerous_configuration` and `secret_extraction` could probably stop being features, and we could make those part of the main API, under a new `::dangerous::` module. Using features to gate dangerous things runs into the classic problem of Cargo features: they are additive, so one of your transitive dependencies could enable them for you, which would allow using the gated items without explicitly enabling them.

For someone auditing codebases there are pros and cons. Pro: putting things in a specific module means you can grep for imports of that module in the codebase (and its dependencies). Con: the feature approach in theory allows an auditor to query the evaluated feature set after dependency resolution and see if `dangerous_configuration` is enabled. I'm not sure which Cargo commands one would use but I'm sure it's possible.

There's also the `tls12` feature. In the config builder we try to make sure unused code (like TLS 1.2) can be eliminated by the linker, but we also have the `tls12` Cargo feature. I haven't dug into how painful it would be to fully remove the feature, but it would have some nice benefits. For instance, we have some code that is only active for `#[cfg(not(feature = "tls12"))]`. I believe that's not understood by our coverage measurements.

### LINKED ISSUE #1472: Crypto provider feature guards are not additive
I have VS Code configured to compile code with `--all-features` by default. After landing #1405 this is showing compilation failures in the new provider-example crate:

* `Tls13CipherSuite` has `pub(crate)` fields that are only enabled for `#[cfg(feature = "quic")]`. This means that other crypto providers can't create `Tls13CipherSuite` values when the `quic` feature is enabled.
* `Tls12AeadAlgorithm` and `Tls13AeadAlgorithm` have `extract_keys()` methods that only appear for `#[cfg(feature = "secret_extraction")]`. This means that also downstream implementations of these traits will also need to guard these trait methods.

This happens because my Rust-Analyzer setup is compiling rustls with all features enabled, but the provider-example crate cannot deal with the additional features -- it can only be compiled when rustls doesn't have these features enabled, but that can never be guaranteed in Cargo's feature-additive dependency graphs.

### LINKED ISSUE #1491: Expose more QUIC mods (and friends)
Although RustCrypto doesn't seems to have QUIC support yet, we should expose the relevant mods so that we at least have a chance to experiment for an implementation.

Currently this includes but not limited to: `rustls::quic`, `rustls::hkdf`.

### PR DESCRIPTION
**Pull request by ctz**
fixes #1491
fixes #1437 
fixes #1472

### CODE REVIEW

**Comment by ctz** [comment]
I think the coverage results are valid; they are quic-specific branches that were previously compiled out of the coverage tests.

File: rustls/src/quic.rs [hunk 1]
```diff
@@ -2,12 +2,16 @@
 use crate::client::{ClientConfig, ClientConnectionData, ServerName};
 use crate::common_state::{CommonState, Protocol, Side};
 use crate::conn::{ConnectionCore, SideData};
-use crate::crypto::tls13::{HkdfExpander, OkmBlock};
+use crate::crypto::cipher::{AeadKey, Iv};
+use crate::crypto::tls13::HkdfExpander;
+use crate::crypto::tls13::OkmBlock;
```
Comment (cpu) [inline rustls/src/quic.rs:7]:
```suggestion
use crate::crypto::tls13::{HkdfExpander, OkmBlock};
```

File: rustls/src/crypto/ring/quic.rs [hunk 2]
```diff
@@ -252,12 +233,16 @@ mod tests {
         let expander = TLS13_CHACHA20_POLY1305_SHA256_INTERNAL
             .hkdf_provider
             .expander_for_okm(&OkmBlock::new(SECRET));
-        let hpk =
-            super::HeaderProtectionKey::new(expander.as_ref(), version, &aead::quic::CHACHA20);
+        let header_key =
+            hkdf_expand_label_aead_key(expander.as_ref(), 32, version.header_key_label(), &[]);
+        let packet_key =
+            hkdf_expand_label_aead_key(expander.as_ref(), 32, version.packet_key_label(), &[]);
```
Comment (cpu) [inline rustls/src/crypto/ring/quic.rs:239]: I think you could knock out the magic 32 constant with a binding that takes advantage of the new `aead_key_len` fn on the trait like:
```
let key_len = TLS13_CHACHA20_POLY1305_SHA256_INTERNAL
            .quic
            .unwrap()
            .aead_key_len();
```

File: connect-tests/Cargo.toml [hunk 3]
```diff
@@ -7,7 +7,6 @@ description = "Rustls connectivity based integration tests."
 publish = false
 
 [features]
```
Comment (cpu) [inline connect-tests/Cargo.toml:9]: nit: maybe drop this empty `[features]` section?

File: examples/Cargo.toml [hunk 4]
```diff
@@ -7,7 +7,6 @@ description = "Rustls example code and tests."
 publish = false
 
 [features]
```
Comment (cpu) [inline examples/Cargo.toml:9]: ditto RE: empty features.

File: rustls/examples/internal/bogo_shim.rs [hunk 5]
```diff
@@ -2,1293 +2,11 @@
 // test suite. See bogo/ for this in action.
 //
 // https://boringssl.googlesource.com/boringssl/+/master/ssl/test
-//
```
Comment (cpu) [inline rustls/examples/internal/bogo_shim.rs:5]: Would it be feasible to separate the `bogo_shim.rs` -> `bogo_shim_impl.rs` move in its own commit ahead of this work, maintaining the feature gates, and then removing those gates in this commit?

File: rustls/examples/internal/bogo_shim.rs [hunk 6]
```diff
@@ -6,6 +6,11 @@
 #[cfg(any(feature = "ring", feature = "aws_lc_rs"))]
 mod bogo_shim_impl;
 
+#[cfg(any(feature = "ring", feature = "aws_lc_rs"))]
+fn main() {
+    bogo_shim_impl::main();
+}
```
Comment (cpu) [inline rustls/examples/internal/bogo_shim.rs]: Should this have been in the previous commit where the move was done and the quic feature removed? :thinking:

File: rustls/examples/internal/bogo_shim_impl.rs [hunk 7]
```diff
@@ -930,7 +930,7 @@ fn exec(opts: &Options, mut sess: Connection, count: usize) {
     }
 }
 
-fn main() {
+pub fn main() {
```
Comment (cpu) [inline rustls/examples/internal/bogo_shim_impl.rs:970]: ditto RE: wondering if this would make sense in the previous commit

File: rustls/src/suites.rs [hunk 8]
```diff
@@ -18,6 +19,22 @@ pub struct CipherSuiteCommon {
 
     /// Which hash function the suite uses.
     pub hash_provider: &'static dyn crypto::hash::Hash,
+
+    /// The number of messages that can be encrypted by a single
+    /// instance of `MessageEncrypter` produced for this suite before
+    /// an attacker gains an advantage in distinguishing it from an ideal PRP.
```
Comment (cpu) [inline rustls/src/suites.rs:25]:
```suggestion
    /// an attacker gains an advantage in distinguishing it from an ideal pseudorandom permutation (PRP).
```

**Review by cpu — COMMENTED** [review]
cool!

File: rustls/examples/internal/bogo_shim.rs (hunk 5 above)
Comment (ctz) [reply rustls/examples/internal/bogo_shim.rs:5]: Yes, will do that.

File: connect-tests/Cargo.toml (hunk 3 above)
Comment (ctz) [reply connect-tests/Cargo.toml:9]: Deleted.

File: examples/Cargo.toml (hunk 4 above)
Comment (ctz) [reply examples/Cargo.toml:9]: Deleted.

File: rustls/src/suites.rs (hunk 8 above)
Comment (ctz) [reply rustls/src/suites.rs:25]: Thanks -- again with the jargon!

**Review by cpu — APPROVED** [review]
:rocket:

File: rustls/src/quic.rs [hunk 9]
```diff
@@ -561,16 +564,26 @@ impl AsRef<[u8]> for Tag {
 }
 
 /// How a `Tls13CipherSuite` generates `PacketKey`s and `HeaderProtectionKey`s.
-pub(crate) trait Algorithm: Send + Sync {
+pub trait Algorithm: Send + Sync {
+    /// Produce a `PacketKey` encrypter/decrypter for this suite.
+    ///
+    /// `suite` is the entire suite this `Algorithm` appeared in.
+    /// `key` and `iv` is the key material to use.
```
Comment (djc) [inline rustls/src/quic.rs:583]: Given that the fields of `AeadKey` are private, I'm not sure this trait is practically implementable for foreign types?

Comment (ctz) [reply rustls/src/quic.rs:583]: The expected things for people to do with an `AeadKey` is:

- view the raw bytes, via `impl AsRef<[u8]>`
- move or drop it (has a non-trivial zeroizing `impl Drop`)

I don't think there is any need for construction of these outside the crate.

Comment (djc) [reply rustls/src/quic.rs:583]: Ahh, yes -- sorry for my confusion.
